Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -661,80 +570,26 @@ std::string Serializer::EncodeQueryPath(const ResourcePath& path) const {
+  return EncodeTimestamp(version.timestamp());
 }
 
-void Serializer::EncodeObjectMap(
-    nanopb::Writer* writer,
-    const model::ObjectValue::Map& object_value_map,
-    uint32_t map_tag,
-    uint32_t key_tag,
-    uint32_t value_tag) {
-  // Write each FieldsEntry (i.e. key-value pair.)
-  for (const auto& kv : object_value_map) {
-    writer->WriteTag({PB_WT_STRING, map_tag});
-    writer->WriteNestedMessage([&kv, &key_tag, &value_tag](Writer* writer) {
-      return EncodeFieldsEntry(writer, kv, key_tag, value_tag);
-    });
-  }
-}
-
-void Serializer::EncodeVersion(nanopb::Writer* writer,
-                               const model::SnapshotVersion& version) {
-  EncodeTimestamp(writer, version.timestamp());
-}
-
-/**
- * Encodes a 'FieldsEntry' object, within a FieldValue's map_value type.
- *
- * In protobuf, maps are implemented as a repeated set of key/values. For
- * instance, this:
- *   message Foo {
- *     map<string, Value> fields = 1;
- *   }
- * would be written (in proto text format) as:
- *   {
- *     fields: {key:"key string 1", value:{<Value message here>}}
- *     fields: {key:"key string 2", value:{<Value message here>}}
- *     ...
- *   }
- *
- * This method writes an individual entry from that list. It is expected that
- * this method will be called once for each entry in the map.
- *
- * @param kv The individual key/value pair to write.
- */
-void Serializer::EncodeFieldsEntry(Writer* writer,
-                                   const ObjectValue::Map::value_type& kv,
-                                   uint32_t key_tag,
-                                   uint32_t value_tag) {
-  // Write the key (string)
-  writer->WriteTag({PB_WT_STRING, key_tag});
-  writer->WriteString(kv.first);
-
-  // Write the value (FieldValue)
-  writer->WriteTag({PB_WT_STRING, value_tag});
-  writer->WriteNestedMessage(
-      [&kv](Writer* writer) { EncodeFieldValue(writer, kv.second); });
+google_protobuf_Timestamp Serializer::EncodeTimestamp(
+    const Timestamp& timestamp_value) {
+  google_protobuf_Timestamp result = google_protobuf_Timestamp_init_zero;
+  result.seconds = timestamp_value.seconds();
+  result.nanos = timestamp_value.nanoseconds();
+  return result;
 }
 
-absl::optional<SnapshotVersion> Serializer::DecodeSnapshotVersion(
-    nanopb::Reader* reader) {
-  absl::optional<Timestamp> version = DecodeTimestamp(reader);
-  if (!reader->status().ok()) return absl::nullopt;
-  return SnapshotVersion{*version};
+SnapshotVersion Serializer::DecodeSnapshotVersion(
+    nanopb::Reader* reader, const google_protobuf_Timestamp& proto) {
+  return SnapshotVersion{DecodeTimestamp(reader, proto)};
 }
 
-absl::optional<Timestamp> Serializer::DecodeTimestamp(nanopb::Reader* reader) {
-  google_protobuf_Timestamp timestamp_proto =
-      google_protobuf_Timestamp_init_zero;
-  reader->ReadNanopbMessage(google_protobuf_Timestamp_fields, &timestamp_proto);
-
+Timestamp Serializer::DecodeTimestamp(
```

Comment:
Is the absl::optional<Timestamp> to Timestamp change required for the change to use malloc-ed nanopb version? If not, could you elaborate the reason for the return type change?

Reply:
No it's not necessary; we could keep it as optional<>. This would additionally have the benefit of forcing the caller to check the status (or die otherwise) whereas as currently written, if the caller forgets to check the status, they'll end up using an invalid return value.

The disadvantage of keeping it as optional<>, is that it introduces a lot more boilerplate code, since now all the methods have to continually check all of their values to ensure a value is present before continuing.

So why the change now? Previously, we did most of the deserialization ourselves, so we were a little more paranoid. Now, nanopb does more of it for us, so we feel a little safer.

Other alternatives that we could consider:
1) use optional<T> for the public methods, and just T for the private ones. The callers get safety, while minimizing (though not eliminating) the internal boilerplate.
2) Similar to (1), but switch to statusor<> rather than optional<>. We've effectively eliminated nanopb/reader and nanopb/writer anyways, so if we merge the last little bits into the serializers, then the public methods can just return statusor<T> while the private methods can continue to return T.